text/0129-video-replay-envelope.md

```diff
@@ -0,0 +1,69 @@
+- Start Date: 2024-02-06
+- RFC Type: feature
+- RFC PR: [#129](https://github.com/getsentry/rfcs/pull/129)
+- RFC Status: draft
+
+# Summary
+
+In order to capture video replays and present them to the user, we need to define/decide:
+
+- how to transport the video data to the server
+- how to integrate the video data into the RRWeb JSON format
+- how to combine multiple replay chunks to a single replay session
+
+All of these influence one another and need to be considered together in a single RFC.
+
+Note: SDK-side implementation that is currently being worked on relies on taking screenshots and encoding them to a video.
+This is based on an evaluation where a video has much smaller size than a sequence of images (**TODO fix these: factor of X for 720p video**).
+
+# Motivation
+
+We need this to to capture replays on platforms where it's not possible/feasible to produce an HTML DOM (i.e. the native format supported by RRWeb). For example: mobile apps.
+
+<!-- # Supporting Data -->
+<!-- Metrics to help support your decision (if applicable). -->
+
+# Options Considered
+
+## Using a video, with EnvelopeItem:ReplayVideo
+
+- From the SDK, we would send a new envelope item type, [`ReplayVideo`](https://github.com/getsentry/relay/blob/5fd3969e88d3eea1f2849e55b61678cac6b14e44/relay-server/src/envelope.rs#L115C5-L115C20) to transport the video data.
+  The envelope item would consist of a single header line (JSON), followed by a new line and the raw video data.
+  - The header should contain at least the following metadata:
+    - the raw size of the video data
+    - the duration of the video in milliseconds
+    - the encoder format
+    - the container format
+    - the video resolution
+  - Additionally, it may be useful to include more information, such as:
+    - the number of frames in the video
+    - the frame rate type (constant, variable)
+    - the frame rate value
```

Comment:
It might make more sense to put these headers (other than segment_id) into the RRWeb JSON.  For example:

```json
{
  "type": 5,
  "data": {
    "type": "video",
    "size": 3440,
    "duration": 5000,
    "encoding": "whatever",
    "container": "whatever",
    "v-size": 1920,
    "h-size": 1080,
    "frame_count": 50,
    "frame_rate_type": "constant",
    "frame_rate": 10
  }
}
```

This simplifies the implementation on the backend and probably the frontend.  We can keep the video blob in its own file and the RRWeb in its own file.  If we put the video metadata into the headers of the video request we would have to either store it in its own file or encode the video to a b64 string and write it into the headers JSON file.  Which would be inefficient.

Reply:
The key names and their values are just placeholders.  Replace them with whatever makes the most sense to the SDK and player implementations.

Reply:
> It might make more sense to put these headers (other than segment_id) into the RRWeb JSON. For example:

The idea to put them into the envelope was to allow us to easily work with the data in relay and elsewhere, without relying on RRWeb JSON format, which may change in the future.